### PR TITLE
Refactor BackoffPolicy implementation.

### PR DESCRIPTION
**Pull request by coryan**

### CODE REVIEW

File: google/cloud/internal/backoff_policy.h [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+   *
+   * Typically implemented as
+   * @code
+   *   return std::unique_ptr<BackoffPolicy>(new Foo(*this));
+   * @endcode
+   */
+  virtual std::unique_ptr<BackoffPolicy> clone() const = 0;
+
+  /**
+   * Return the delay after an RPC operation has completed.
```
Comment (houglum) [inline google/cloud/internal/backoff_policy.h]: Suggested docstring change --> "Return the delay (in ms) to wait before the next retry attempt."

File: google/cloud/internal/backoff_policy.h [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+   *     documented in `std::chrono::duration<>` (in brief, the length of the
+   *     tick in seconds, expressed as a `std::ratio<>`), for our purposes it
+   *     is simply a formal parameter.
+   * @tparam Period2 similar formal parameter for the type of @p maximum_delay.
+   *
+   * @see
+   * [std::chrono::duration<>](http://en.cppreference.com/w/cpp/chrono/duration)
+   *     for more details.
+   */
+
```
Comment (houglum) [inline google/cloud/internal/backoff_policy.h]: Nit: remove extra line

File: google/cloud/internal/backoff_policy.h [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+   *     operation.
+   * @param maximum_delay the maximum value for the delay between operations.
+   * @param scaling how fast does the delay increase between iterations.
+   *
+   * @tparam Rep1 a placeholder to match the Rep tparam for @p initial_delay's
+   *     type, the semantics of this template parameter are documented in
+   *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
+   *     used to store the number of ticks), for our purposes it is simply a
+   *     formal parameter.
+   * @tparam Rep2 similar formal parameter for the type of @p maximum_delay.
```
Comment (houglum) [inline google/cloud/internal/backoff_policy.h]: Nit: These are documented in a different order than they're defined below.

File: google/cloud/internal/backoff_policy.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/backoff_policy.h]: Generally I like this, with a few nits:

- The 'Return' thing is already put [there](https://googlecloudplatform.github.io/google-cloud-cpp/0.1.0/bigtable/classbigtable_1_1v0_1_1ExponentialBackoffPolicy.html#a48e212f5e83c54fc19630afb63d7ef14) by Doxygen
- The return type is called `std::chrono::milliseconds` is it worth it to repeat the units?  I will change it if you think it is.

File: google/cloud/internal/backoff_policy.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/backoff_policy.h]: Done.

File: google/cloud/internal/backoff_policy.h (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/backoff_policy.h]: Fixed.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/internal/backoff_policy.h [hunk 4]
```diff
@@ -58,9 +58,14 @@ class BackoffPolicy {
   virtual std::unique_ptr<BackoffPolicy> clone() const = 0;
 
   /**
-   * Return the delay after an RPC operation has completed.
+   * Handle an operation completion.
    *
-   * @return how long to wait before trying the operation again.
+   * This function is typically called when an operation has failed (if it had
```
Comment (houglum) [inline google/cloud/internal/backoff_policy.h]: Nit: This is a big run-on sentence now.  Switch out all the commas for periods, please :)

Also, the doc summary is much nicer now. Thanks!

File: google/cloud/internal/backoff_policy.h [hunk 5]
```diff
@@ -103,19 +108,18 @@ class ExponentialBackoffPolicy : public BackoffPolicy {
...
    *     used to store the number of ticks), for our purposes it is simply a
    *     formal parameter.
-   * @tparam Rep2 similar formal parameter for the type of @p maximum_delay.
    * @tparam Period1 a placeholder to match the Period tparam for
    *     @p initial_delay's type, the semantics of this template parameter are
    *     documented in `std::chrono::duration<>` (in brief, the length of the
    *     tick in seconds, expressed as a `std::ratio<>`), for our purposes it
    *     is simply a formal parameter.
    * @tparam Period2 similar formal parameter for the type of @p maximum_delay.
+   * @tparam Rep2 similar formal parameter for the type of @p maximum_delay.
```
Comment (houglum) [inline google/cloud/internal/backoff_policy.h]: These are still out of order with the params below (switch doc placement for Period2 and Rep2).

File: google/cloud/internal/backoff_policy.h (hunk 4 above)
Comment (coryan) [reply google/cloud/internal/backoff_policy.h]: Fixed.

File: google/cloud/internal/backoff_policy.h (hunk 5 above)
Comment (coryan) [reply google/cloud/internal/backoff_policy.h]: Fixed.

**Pull request merged**